src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs

```diff
@@ -99,15 +99,7 @@ public override void Write(string value)
         /// <inheritdoc />
         public override void Write(IHtmlContent value)
         {
-            var htmlTextWriter = TargetWriter as HtmlTextWriter;
-            if (htmlTextWriter == null)
-            {
-                value.WriteTo(TargetWriter, HtmlEncoder);
-            }
-            else
-            {
-                htmlTextWriter.Write(value);
-            }
+            value.WriteTo(TargetWriter, HtmlEncoder);
```

Comment:
We need to remove this optimization and put it back into the places where it's valuable. 

The reason is that often see a value show up here like: `TagHelperOutput`->`TagHelperContent`->`ViewBuffer`. We want 'flatten' this hierarchy, but without any hardcoding to types like `TagHelperOutput`. 

The solution is to remove this check, and let `TagHelperOutput` control its own behavior.

This ultimately plays much nicer if data gets copied into the `ViewBuffer`. That lets intermediate lists get GCed before they can graduate from gen0. `ViewBuffer` is pooled so there's no such issue.